_posts/2018-10-05-a-macroscopic-take-on-metaprogramming.md

```diff
@@ -0,0 +1,293 @@
+- or the mock object
+
+Another way to look at this problem would be to wonder if we could hijack this `_getDispatcher` function and mock it directly at the "source".
+
+es6 proxies allow just that.
+They offer a way to catch just what you need of the behavior or add to it.
+
+So in our example, we replace the mock with the below:
+```
+const FirstExternalApiMock = new Proxy(FirstExternalApi, {
+  get(target, propKey) {
+    if (propKey === '_getDispatcher') {
+      return () => {
+        return (method, ...args)=> {
+          const argsString = args.map(arg => {
+            return JSON.stringify(arg, null, 4);
+        }).join(', ');
+
+          console.log(`externalApi: ('${argsString}')`);
+        }
+      };
+    } else {
+      return target[propKey];
+    }
+  }
+});
+```
+
+We pass the object we are interested in.
+'get' is responsible for 'listening' to whenever a property of this object is called.
+Now we leave all properties but one alone and focus entirely on 'trapping' `_getDispatcher`.
+When it's called, we output the arguments to the console.
+
+It's a different, slightly more concise more way of implementing our solution.
+
+## Elixir
+
+In Elixir, most of the metaprogramming boils down to understanding the functions "quote" and "unquote" (by no means a measure of simplicity).
+Quote effectively gives us access to Elixir's AST (abstract syntax tree) by showing us how an expression is parsed and transformed into a tree that Elixir can handle.
+What is returned typically contains three-element tuples (operation/data type, metadata, arguments/data).
+
+### Example
+
+Again, we would like to make our test suite a bit more concise.
+Our application offers two endpoints with a very similar set of functionality.
+This should be a good case for abstracting most of their testing and sharing the tests between the two endpoints (endpoint_a and endpoint_b).
+Not dissimilar to RSpec shared example feature, making use of a domain specific language would work for our purposes.
+
+We have a series of tests like the one below:
+
+```
+defmodule Api.V1.AControllerTest do
+	test "#create http returns a 403 status with json error" do
+	  "/v1/endpoint_a"
+	  |> create_record(403, ssl_error())
+	end
+end
+
+defmodule Api.V1.BControllerTest do
+	test "#create http returns a 403 status with json error" do
+	  "/v1/endpoint_b"
+	  |> create_record(403, ssl_error())
+	end
+end
+
+```
+
+The only difference being the endpoint, we can abstract this to:
+
+```
+defmodule SharedTestCase do
+  use ExUnit.Case
+
+  @moduletag endpoint: "endpoint_a"
+
+  test "#create http returns a 403 status with json error" do
+  		"/v1/#{endpoint}"
+  		|> create_record(403, ssl_error())
+	end
+end
+```
+
+@moduletag is a module attribute provided by ExUnit: one of its uses being to act as a temporary module storage to be used during compilation.
+
+A macro is going to help Api.V1.AControllerTest and Api.V1.BControllerTest modules make use of the SharedTestCase module.
```

Comment:
Especially given the "Macro"-scopic view of this post, I'd introduce macros earlier in this section, and with a bit more depth. Macros **are** after all the pinnacle of metaprogramming... Well at least until the AIs are writing my code.

Reply:
@nixpulvis Ok, it's hard to know when to stop to be honest but in this case even more so. I did skip talking about `eigenclasses` even though I think they are important to the concept of metaprogramming in Ruby just because I wanted to tackle 3 practical real-life examples in different languages and I was struggling to make a short point about them. My point is, I am sure you have an idea of what I missed that you would have included and that would massively help me polish up the post. Thanks!

Reply:
You start talking about Elixir macros specifically before introducing the concept. Sometimes leading the way with examples is the best way to go, but here I feel a short paragraph about what a macro really _is_, is in order.

If this helps you, here's how I think about macros and meta-programming a bit. There are two kinds of abstraction, dynamic and static. There are also two constraints for any program, it must uphold the correct semantics of your desired behavior (a `sort` function should return a sorted list), and it must be syntactically valid.

A dynamic abstraction (function, method, etc) can be used to satisfy the semantics of your program, but it's up to you to correctly construct the syntax.

A static abstraction (macro) can be used to satisfy both the semantics **and** syntax of a program. But only does so by creating a new set of syntax and semantics to use.

So the question is *really* what kinds of problems are macros, or meta programming in general good for, and that's something I feel the rest of this piece already does a pretty good job motivating. **Generally though, *macros are for operating on syntax, while functions are for operating on data**, if you want to DRY up some code that is only syntactically similar, a macro might be a good choice.

As you use more and more macros the distinction between what is static or dynamic starts to blur, and you can end up with some very interesting semantics. Racket (and LISP originally) for example is basically a big collection of macros used to create a programming language. Implementing a [lazy](https://docs.racket-lang.org/lazy/index.html) mode for the language wasn't too hard. In fact a compiler can be thought of as a big fancy macro.

I was writing a few macros over the weekend, and they also felt like great examples, maybe you'll get a kick out of these: https://github.com/nixpulvis/oursh/blob/master/tests/piped.rs.

Reply:
@nixpulvis As mentioned on Slack, it would be great to use this as a starting point for a follow-up blog post! The blog does need posts 👍